llvm/test/MC/AMDGPU/hsa-sym-exprs-gfx11.s

```diff
@@ -117,18 +119,19 @@ expr_defined:
-// ASM-NEXT: .amdhsa_fp16_overflow (((((((((((((((((((((((((((((((0&(~786432))|(3<<18))&(~2097152))|(1<<21))&(~8388608))|(1<<23))&(~536870912))|(1<<29))&(~1073741824))|(1<<30))&(~12288))|(defined_2_bits<<12))&(~49152))|(defined_2_bits<<14))&(~196608))|(defined_2_bits<<16))&(~786432))|(defined_2_bits<<18))&(~67108864))|(defined_boolean<<26))&(~536870912))|(defined_boolean<<29))&(~1073741824))|(defined_boolean<<30))&(~2147483648))|(defined_boolean<<31))&(~63))|(0<<0))&(~960))|(0<<6))&67108864)>>26
-// ASM-NEXT: .amdhsa_workgroup_processor_mode (((((((((((((((((((((((((((((((0&(~786432))|(3<<18))&(~2097152))|(1<<21))&(~8388608))|(1<<23))&(~536870912))|(1<<29))&(~1073741824))|(1<<30))&(~12288))|(defined_2_bits<<12))&(~49152))|(defined_2_bits<<14))&(~196608))|(defined_2_bits<<16))&(~786432))|(defined_2_bits<<18))&(~67108864))|(defined_boolean<<26))&(~536870912))|(defined_boolean<<29))&(~1073741824))|(defined_boolean<<30))&(~2147483648))|(defined_boolean<<31))&(~63))|(0<<0))&(~960))|(0<<6))&536870912)>>29
-// ASM-NEXT: .amdhsa_memory_ordered (((((((((((((((((((((((((((((((0&(~786432))|(3<<18))&(~2097152))|(1<<21))&(~8388608))|(1<<23))&(~536870912))|(1<<29))&(~1073741824))|(1<<30))&(~12288))|(defined_2_bits<<12))&(~49152))|(defined_2_bits<<14))&(~196608))|(defined_2_bits<<16))&(~786432))|(defined_2_bits<<18))&(~67108864))|(defined_boolean<<26))&(~536870912))|(defined_boolean<<29))&(~1073741824))|(defined_boolean<<30))&(~2147483648))|(defined_boolean<<31))&(~63))|(0<<0))&(~960))|(0<<6))&1073741824)>>30
-// ASM-NEXT: .amdhsa_forward_progress (((((((((((((((((((((((((((((((0&(~786432))|(3<<18))&(~2097152))|(1<<21))&(~8388608))|(1<<23))&(~536870912))|(1<<29))&(~1073741824))|(1<<30))&(~12288))|(defined_2_bits<<12))&(~49152))|(defined_2_bits<<14))&(~196608))|(defined_2_bits<<16))&(~786432))|(defined_2_bits<<18))&(~67108864))|(defined_boolean<<26))&(~536870912))|(defined_boolean<<29))&(~1073741824))|(defined_boolean<<30))&(~2147483648))|(defined_boolean<<31))&(~63))|(0<<0))&(~960))|(0<<6))&2147483648)>>31
+// ASM-NEXT: .amdhsa_next_free_vgpr defined_value+4
+// ASM-NEXT: .amdhsa_next_free_sgpr defined_value+5
+// ASM-NEXT: .amdhsa_reserve_vcc defined_boolean
+// ASM-NEXT: .amdhsa_float_round_mode_32 (((((((((((((((((((((((((((((((0&(~786432))|(3<<18))&(~2097152))|(1<<21))&(~8388608))|(1<<23))&(~536870912))|(1<<29))&(~1073741824))|(1<<30))&(~12288))|(defined_2_bits<<12))&(~49152))|(defined_2_bits<<14))&(~196608))|(defined_2_bits<<16))&(~786432))|(defined_2_bits<<18))&(~67108864))|(defined_boolean<<26))&(~536870912))|(defined_boolean<<29))&(~1073741824))|(defined_boolean<<30))&(~2147483648))|(defined_boolean<<31))&(~63))|((((alignto(max(defined_value+4, 1), 8))/8)-1)<<0))&(~960))|((((alignto(max(0, 1), 8))/8)-1)<<6))&12288)>>12
```

Comment:
This may be coming a bit late, as the patch where these kinds of expressions were introduced is already landed, but is there a way to simplify these? We control how these pseudo-directives are defined, so it is at least feasible, right?

I know that hand-written assembler is "for experts", and that in this form it can act more as a symbolic relocatable, but one can end up with this form of build-the-whole-descriptor-and-then-locate-the-relevant-field expression in lieu of an error from something as trivial as a typo now, and when going to an object file you get no hints as to where the real root of the issue lies:

```
// whoops, meant the absolute value 0 here:
.amdhsa_next_free_sgpr a0

$ llvm-mc foo.s
<some unintelligible mess>
$ llvm-mc -filetype=obj foo.s
<unknown>:0: error: expected relocatable expression
```

Maybe we don't need to do anything, but it just feels "wrong" as-is.

Reply:
Thanks for comments!

I'd argue simplification is possible, I do try resolving any MCExprs where possible before having to fall back to printing the monolithic expression but since a lot of them are derived from composed OR'ed registers (e.g., `COMPUTE_PGM_RSRC1/2/3`) *any* of the bits that depend on not-yet-resolvable MCExprs means the *whole* MCExpr needs to be printed. Feels unnecessary if only 2 bits are to be printed and even more so if said 2 bits printed are resolvable. I'm thinking either explicitly splitting up the composites of the OR'ed registers and tracking/printing them separately, or having some simplification function that can extract only the sub-expr relevant to what we're trying to check/print/whatever.

The error comment resonates with me: some of the MCExpr-fied variables can now not be validated at the same time as before as they may be unresolvable MCExprs at the time they would normally be validated. With some of them I've tried to delay validation but that is no assurance and additionally, that may mean a loss in error message granularity as delaying it may result in a loss of asm source code location information. I haven't been able to get a satisfactory solution for this.

Let me know if I'm missing some kind of low hanging fruit, though. Would love to hear your thoughts on this.